Multiple reader/writer api

rhpvorderman opened the pull request:
I added the ability for `dnaio.open` to open multiple files.  I also added the `dnaio.open_multiple` API which iworks in case you want 1-tuples for 1 file. (I use this sort of code in my fastq filter and deduplication programs)

dnaio.open is backwards compatible. Also I think `MultipleFastqWriter` is probably going to be faster than `PairedFastqWriter`, but I will benchmark that later and backport any changes.

The `MultipleFastXWriter` objects also have a `write_iterable` call which you can give a SequenceRecord tuple generator. This is much faster than repeatedly calling write calls in a for loop. This can also be backported to the PairedFastqWriter and SingleFastqWriter at a later date.

The API is fully tested. Just need to add some docstrings. Posting this here in advance for code review and to make sure this effort is not duplicated.

rhpvorderman commented:
on second thought, this extra open_multiple API is going to be confusing and there is some code duplication. Let me simplify this.

marcelm commented:
Nice, I’ve started a review and have gotten halfway, will submit when I’m finished.

marcelm commented on src/dnaio/multipleend.py:
```diff
@@ -0,0 +1,205 @@
+        for record, writer in zip(records, self.writers):
+            writer.write(record.fastq_bytes())
+
+    def write_iterable(self, records_iterable: Iterable[Tuple[SequenceRecord, ...]]):
```
This needs a docstring that describes when to use this method over `write()`.

I’m not so fond of the name; would something like `write_many` be clearer? (Not perfect either, though.)

rhpvorderman replied:
write_multiple? 
write_list?
Write list sounds quite okay, and properly conveys that the list can also have one entry. That it accepts any iterable container type should not really be that important.

marcelm replied:
Hm, I think `write_iterable` is still better than those alternatives. Perhaps the method just needs a docstring and then it’s fine.

rhpvorderman replied:
The method has a docstring trough the MultipleFileWriter interface. It is an abstractmethod that is inherited.

marcelm commented on tests/test_multiple.py:
```diff
@@ -0,0 +1,121 @@
+    ["number_of_files", "fileformat"],
+    itertools.product((1, 2, 3, 4), ("fastq", "fasta")),
+)
+def test_multiple_write_too_much(number_of_files, fileformat):
```
Should be `_too_many`

marcelm reviewed (commented):
I think this is a good addition. I am a little bit concerned, though, that this is code that will see much less use (and therefore testing) than the code for the 1- and 2-file cases because there’s presumably very little data out there that needs this. What would be the downside of not only backporting some of this to the PairedEnd... classes, but merging them (at some point in the future, not now) so that there’s only one way to read more than one file?

A note regarding terminology: I noticed in the code that the variable name "records" was used quite often, but it’s now less clear whether that refers to records belonging to the same molecule fragment (forward and reverse read in the case of paired-end reads) or whether it refers to multiple unrelated single ends. That is, if you imagine the data as a table, "records" could refer to a row or a column. I’m not sure what to do about this. Perhaps it’s clear enough from the context.

marcelm commented:
> What would be the downside of not only backporting some of this to the PairedEnd... classes, but merging them (at some point in the future, not now) so that there’s only one way to read more than one file?

It appears I forgot the discussion we had in #82.

rhpvorderman commented on src/dnaio/multipleend.py:
```diff
@@ -0,0 +1,205 @@
+        raise ValueError("mode must be one of 'r', 'w', 'a'")
+    if mode == "r":
+        return MultipleFileReader(*files, fileformat=fileformat, opener=opener)
+    append = True if mode == "a" else False
```
:man_facepalming:

rhpvorderman commented on src/dnaio/multipleend.py:
```diff
@@ -0,0 +1,205 @@
+    if mode == "r":
+        return MultipleFileReader(*files, fileformat=fileformat, opener=opener)
+    append = True if mode == "a" else False
+    if fileformat == "fastq" or qualities or (fileformat is None and qualities is None):
```
I added format detection. I assume the first file has the same type as the others. I will bet a beer (or other beverage) on it that this assumption will not lead to an issue on our bug tracker in the coming 10 years.
No need to add extra code to catch every corner case.

rhpvorderman commented on src/dnaio/__init__.py:
```diff
@@ -122,11 +122,20 @@ def open(
-    elif interleaved or file2 is not None:
+    if mode not in ("r", "w", "a"):
+        raise ValueError("Mode must be 'r', 'w' or 'a'")
+    if files and file2 is not None:
```
Agreed. Fixed it.

rhpvorderman commented on src/dnaio/multipleend.py:
```diff
@@ -0,0 +1,205 @@
+            raise ValueError("At least one file is required")
+        self.files = files
+        self.readers: List[Union[FastaReader, FastqReader]] = [
+            _open_single(  # type: ignore
```
Good catch. Implemented

rhpvorderman commented on src/dnaio/__init__.py:
```diff
@@ -69,7 +70,7 @@
     file1: Union[str, PathLike, BinaryIO],
-    *,
+    *files: Union[str, PathLike, BinaryIO],
     file2: Optional[Union[str, PathLike, BinaryIO]] = None,
```
technically `file1` should also be deprecated. This will make the interface much easier ->
```python3
dnaio.open("my.fastq", **kwargs) -> SingleEnd
dnaio.open("R1.fastq", "R2.fastq", **kwargs) -> PairedEnd
dnaio.open("R1.fastq", "R2.fastq", "R3.fastq", **kwargs) -> MultipleEnd
```
It will also become easier for us to use typing.overload and provide the correct typing information.

But in order to be backwards-compatible, file1 and file2 are kept intact.

rhpvorderman replied:
Hmm. That is a good idea actually. I will remake the interface so that `*files` is the primary way of entring files. File1 and file2 should be deprecated keyword arguments. I will change the logic to accomodate this. This will also make it much easier to write the docstring.

marcelm replied:
I like that!

rhpvorderman commented:
Done! Sorry for letting this one rest a bit longer. We are changing a clinical pipeline in our institute and that takes up most of my time.
Also ironically our sequence center has decided to use new machines/software where instead of providing an extra UMI sequence file, they can put it in the header. So this PR suddenly got less relevant. On the other hand, we do have `3` hardcoded multiple times in our code because it is a thing.

Documenting the *files parameter was hard until I decided to also deprecate `file1`. Now it is very straight forward. I just needed to add an elif tree to ensure backwards compatibility but at least the interface should be clear going forward. (If you agree, that is.)

rhpvorderman commented on src/dnaio/__init__.py:
```diff
@@ -122,21 +131,51 @@ def open(
-            file1,
-            file2=file2,
+            files[0],
+            file2=None if interleaved else files[1],
```
I see some redundancy here in open paired. Either you give one or two files. One means interleaved, two means not. The interleaved parameter is this not necessary for the internal _open_paired function. It will be much easier simply to use `_open_paired(*files, **other_kwargs` since we have  already established above that the correct conditions are met.

marcelm replied:
I agree, this could be simpler.

marcelm commented on src/dnaio/__init__.py:
```diff
@@ -68,31 +69,39 @@
 
     Parameters:
+      files:
+        one or multiple Path or an open file-like objects. One for single-end
+        reads, two for multiple-end reads etc. More than two files are also
```
```suggestion
        one or more Path or open file-like objects. One for single-end
        reads, two for paired-end reads etc. More than two files are also
```

marcelm commented on src/dnaio/__init__.py:
```diff
@@ -68,31 +69,39 @@
       mode:
-        Either ``'r'`` or ``'rb'`` for reading, ``'w'`` for writing
-        or ``'a'`` for appending.
+        Either ``'r'``, ``'w'`` for writing or ``'a'`` for appending.
```
```suggestion
        Set to ``'r'`` for reading, ``'w'`` for writing or ``'a'`` for appending.
```

marcelm commented on src/dnaio/__init__.py:
```diff
@@ -122,21 +131,51 @@ def open(
-    elif interleaved or file2 is not None:
+    if files and (file1 is not None):
+        raise ValueError(
+            "the file1 argument cannot be used together with unnamed files"
```
```suggestion
            "The file1 keyword argument cannot be used together with files specified"
            "as positional arguments"
```
(I initially associated "unnamed file" with something filesystem related.)

marcelm commented on src/dnaio/__init__.py:
```diff
@@ -122,21 +131,51 @@ def open(
+    elif len(files) > 1 and file2 is not None:
+        raise ValueError(
+            "the file2 argument cannot be used together with more than one "
+            "unnamed file"
```
```suggestion
            "file specified as positional argument"
```

marcelm commented on src/dnaio/__init__.py:
```diff
@@ -122,21 +131,51 @@ def open(
+        )
+    elif file1 is not None and file2 is not None and files:
+        raise ValueError(
+            "file1 and file2 arguments cannot be used together with unnamed files"
```
```suggestion
            "file1 and file2 arguments cannot be used together with files specified"
            "as positional arguments"
```

marcelm commented on src/dnaio/__init__.py:
```diff
@@ -122,21 +131,51 @@ def open(
+
+    if len(files) > 1 and interleaved:
+        raise ValueError(
+            "When interleaved is True only one file must be specified as input."
```
```suggestion
            "When interleaved is True, only one file must be specified."
```

marcelm commented on src/dnaio/__init__.py:
```diff
@@ -122,21 +131,51 @@ def open(
             mode=mode,
             qualities=qualities,
         )
+    elif len(files) > 2:  # 3 or more files
```
```suggestion
    elif len(files) > 2:
```

marcelm commented on src/dnaio/interfaces.py:
```diff
@@ -40,3 +43,28 @@ def write(self, record1: SequenceRecord, record2: SequenceRecord) -> None:
+    @abstractmethod
+    def write_iterable(self, list_of_records: Iterable[Tuple[SequenceRecord, ...]]):
+        """
+        Iterates over the list (or other iterable container) and writes all
```
```suggestion
        Iterate over the list (or other iterable container) and write all
```

marcelm commented on src/dnaio/__init__.py:
```diff
@@ -122,21 +131,51 @@ def open(
+        )
+    elif len(files) > 1 and file2 is not None:
+        raise ValueError(
+            "the file2 argument cannot be used together with more than one "
```
```suggestion
            "The file2 argument cannot be used together with more than one "
```

marcelm commented on src/dnaio/multipleend.py:
```diff
@@ -0,0 +1,221 @@
+    opener=xopen,
+):
+    if not files:
+        raise ValueError("at least one file is required")
```
```suggestion
        raise ValueError("At least one file is required")
```

marcelm commented on src/dnaio/multipleend.py:
```diff
@@ -0,0 +1,221 @@
+    if not files:
+        raise ValueError("at least one file is required")
+    if mode not in ("r", "w", "a"):
+        raise ValueError("mode must be one of 'r', 'w', 'a'")
```
```suggestion
        raise ValueError("Mode must be one of 'r', 'w', 'a'")
```

marcelm commented on src/dnaio/multipleend.py:
```diff
@@ -0,0 +1,221 @@
+    elif mode == "w" and fileformat is None:
+        # Assume mixed files will not be offered.
+        for file in files:
+            if isinstance(file, str):
```
Should this also be made to work if `file` is `PathLike`?

rhpvorderman replied:
That is quite an easy thing. Done.

marcelm commented on tests/test_open.py:
```diff
@@ -301,3 +302,50 @@ def test_islice_gzip_does_not_fail(tmp_path):
+def test_no_file2_with_multiple_args():
+    with pytest.raises(ValueError) as error:
+        _ = dnaio.open(os.devnull, os.devnull, file2=os.devnull)
+    error.match("more than one unnamed file")
```
Also needs to be adjusted if you accept my suggestion about changing the error message

marcelm reviewed (commented):
Sorry about the very long delay. Are you still interested in merging this? I think your most recent change of deprecating the `file1` and `file2` keyword arguments and instead just having a `*files` positional argument is a very nice improvement.

I have added a couple of suggestions, but they are mostly nitpicks about spelling.

rhpvorderman commented:
> Sorry about the very long delay. Are you still interested in merging this? 

Yes. Aren't you? :wink: https://github.com/marcelm/cutadapt/issues/491#issuecomment-1274754051 
I have two applications (fastq-filter and fastqdedup) where I run into this issue with more than 2 FASTQ files per readgroup. I still think it is nice to have dnaio solve this for me rather than adding custom code for every project.

> I think your most recent change of deprecating the file1 and file2 keyword arguments and instead just having a *files positional argument is a very nice improvement.

It is really weird. Sometimes programming for the general case is an excercise in pedantry which leads to byzantine and complex code, and in other cases it makes the API simpler. Bizarre. Is there a computer science theorem that tackles this problem? (I am a biologist by training).

I have committed al the nitpicks and simplified _open_paired. I also fixed an annoying mypy issue related to `sequence_class`.

marcelm commented:
> > Sorry about the very long delay. Are you still interested in merging this?
> 
> Yes. Aren't you? wink [marcelm/cutadapt#491 (comment)](https://github.com/marcelm/cutadapt/issues/491#issuecomment-1274754051) I have two applications (fastq-filter and fastqdedup) where I run into this issue with more than 2 FASTQ files per readgroup. I still think it is nice to have dnaio solve this for me rather than adding custom code for every project.

Absolutely, I’m just not sure I’ll be able to work on that issue in the foreseeable future. But this PR is the prerequisite for doing it at all.

> Bizarre. Is there a computer science theorem that tackles this problem? (I am a biologist by training).

I thought I heard something like this in a computer science context, but googling now only [gives me some results in mathematics](https://mathoverflow.net/questions/21214/). But this appears to be a thing, yes.

> I have committed al the nitpicks and simplified _open_paired. I also fixed an annoying mypy issue related to `sequence_class`.

I realized now that we need to make sure that the documentation covers the new classes as well. Do you want to do this here or in a separate PR? I can also help.

rhpvorderman commented:
> I thought I heard something like this in a computer science context, but googling now only [gives me some results in mathematics](https://mathoverflow.net/questions/21214/). But this appears to be a thing, yes.

Interesting. Thanks for the link.

> I realized now that we need to make sure that the documentation covers the new classes as well. Do you want to do this here or in a separate PR? I can also help.

I have added it just now. What should I do with `MultipleFileReader`? It is an interface obviously, but also, there is only one class that uses the interface. So I now have it documented as a plain class. That seemed to be the least convoluted way to solve this.

marcelm commented:
> What should I do with MultipleFileReader? It is an interface obviously, but also, there is only one class that uses the interface. So I now have it documented as a plain class. That seemed to be the least convoluted way to solve this.

I don’t have an opinion at the moment, let’s try it this way.

I think this is ready, let’s merge it!

marcelm commented:
I changed a setting on RTD to let it build the documentation on PRs. And I invited you as a co-maintainer.

rhpvorderman commented:
> I changed a setting on RTD to let it build the documentation on PRs. And I invited you as a co-maintainer.

Nice, that should keep the docstrings tidy. I accepted the invitation, thanks!
